[terminal] add link matcher for local files.

svenefftinge opened the pull request:
This PR adds link matching for local files to the terminal.

svenefftinge commented on packages/terminal/src/browser/terminal-linkmatcher-files.ts:
```diff
@@ -0,0 +1,172 @@
+    }
+}
+
+// below reg exps are taken from
```
https://dev.eclipse.org/ipzilla/show_bug.cgi?id=19821

AlexTugarev replied:
@svenefftinge, I think we need to attach an archive containing the specific revision of the file copied from and the LICENSE.
If you want, I'll add it to the CQ

svenefftinge replied:
Ok, thank you.

kittaakos commented on packages/terminal/src/browser/terminal-linkmatcher-files.ts:
```diff
@@ -0,0 +1,172 @@
+            if (toOpen) {
+                // TODO: would be better to ask the opener service, but it returns positively even for unknown files.
+                const f = await this.fileSystem.getFileStat(toOpen.toString());
+                return f !== undefined && !f.isDirectory;
```
Strange, this is `null` on Windows.

```
terminal-linkmatcher-files.ts:54 Uncaught (in promise) TypeError: Cannot read property 'isDirectory' of null
    at TerminalLinkmatcherFiles.<anonymous> (terminal-linkmatcher-files.ts:54)
    at step (terminal-linkmatcher-files.ts:15)
    at Object.next (terminal-linkmatcher-files.ts:15)
    at fulfilled (terminal-linkmatcher-files.ts:15)
(anonymous) @ terminal-linkmatcher-files.ts:54
step @ terminal-linkmatcher-files.ts:15
(anonymous) @ terminal-linkmatcher-files.ts:15
fulfilled @ terminal-linkmatcher-files.ts:15
Promise.then (async)
step @ terminal-linkmatcher.ts:15
(anonymous) @ terminal-linkmatcher.ts:15
push.../../packages/terminal/lib/browser/terminal-linkmatcher.js.__awaiter @ terminal-linkmatcher.ts:15
validationCallback @ terminal-linkmatcher.ts:53
_loop_1 @ Linkifier.ts:246
push.../../node_modules/xterm/lib/Linkifier.js.Linkifier._doLinkifyRow @ Linkifier.ts:207
push.../../node_modules/xterm/lib/Linkifier.js.Linkifier._linkifyRows @ Linkifier.ts:122
(anonymous) @ Linkifier.ts:87
setTimeout (async)
push.../../node_modules/xterm/lib/Linkifier.js.Linkifier.linkifyRows @ Linkifier.ts:87
push.../../node_modules/xterm/lib/Terminal.js.Terminal._queueLinkification @ Terminal.ts:1215
(anonymous) @ Terminal.ts:688
push.../../node_modules/xterm/lib/common/EventEmitter2.js.EventEmitter2.fire @ EventEmitter2.ts:46
(anonymous) @ Terminal.ts:838
push.../../node_modules/xterm/lib/common/EventEmitter2.js.EventEmitter2.fire @ EventEmitter2.ts:46
push.../../node_modules/xterm/lib/renderer/Renderer.js.Renderer._renderRows @ Renderer.ts:206
push.../../node_modules/xterm/lib/ui/RenderDebouncer.js.RenderDebouncer._innerRefresh @ RenderDebouncer.ts:56
(anonymous) @ RenderDebouncer.ts:42
requestAnimationFrame (async)
push.../../node_modules/xterm/lib/ui/RenderDebouncer.js.RenderDebouncer.refresh @ RenderDebouncer.ts:42
push.../../node_modules/xterm/lib/renderer/Renderer.js.Renderer.refreshRows @ Renderer.ts:197
push.../../node_modules/xterm/lib/Terminal.js.Terminal.refresh @ Terminal.ts:1204
push.../../node_modules/xterm/lib/renderer/CursorRenderLayer.js.CursorRenderLayer.onFocus @ CursorRenderLayer.ts:79
(anonymous) @ Renderer.ts:158
(anonymous) @ Renderer.ts:182
push.../../node_modules/xterm/lib/renderer/Renderer.js.Renderer._runOperation @ Renderer.ts:182
push.../../node_modules/xterm/lib/renderer/Renderer.js.Renderer.onFocus @ Renderer.ts:158
(anonymous) @ Terminal.ts:776
push.../../node_modules/xterm/lib/common/EventEmitter.js.EventEmitter.emit @ EventEmitter.ts:73
push.../../node_modules/xterm/lib/Terminal.js.Terminal._onTextAreaFocus @ Terminal.ts:582
(anonymous) @ Terminal.ts:752
terminal-linkmatcher-files.ts:54 Uncaught (in promise) TypeError: Cannot read property 'isDirectory' of null
    at TerminalLinkmatcherFiles.<anonymous> (terminal-linkmatcher-files.ts:54)
    at step (terminal-linkmatcher-files.ts:15)
    at Object.next (terminal-linkmatcher-files.ts:15)
    at fulfilled (terminal-linkmatcher-files.ts:15)
(anonymous) @ terminal-linkmatcher-files.ts:54
step @ terminal-linkmatcher-files.ts:15
(anonymous) @ terminal-linkmatcher-files.ts:15
fulfilled @ terminal-linkmatcher-files.ts:15
Promise.then (async)
step @ terminal-linkmatcher.ts:15
(anonymous) @ terminal-linkmatcher.ts:15
push.../../packages/terminal/lib/browser/terminal-linkmatcher.js.__awaiter @ terminal-linkmatcher.ts:15
validationCallback @ terminal-linkmatcher.ts:53
_loop_1 @ Linkifier.ts:246
push.../../node_modules/xterm/lib/Linkifier.js.Linkifier._doLinkifyRow @ Linkifier.ts:207
push.../../node_modules/xterm/lib/Linkifier.js.Linkifier._linkifyRows @ Linkifier.ts:122
(anonymous) @ Linkifier.ts:87
setTimeout (async)
push.../../node_modules/xterm/lib/Linkifier.js.Linkifier.linkifyRows @ Linkifier.ts:87
push.../../node_modules/xterm/lib/Terminal.js.Terminal._queueLinkification @ Terminal.ts:1215
(anonymous) @ Terminal.ts:688
push.../../node_modules/xterm/lib/common/EventEmitter2.js.EventEmitter2.fire @ EventEmitter2.ts:46
(anonymous) @ Terminal.ts:838
push.../../node_modules/xterm/lib/common/EventEmitter2.js.EventEmitter2.fire @ EventEmitter2.ts:46
push.../../node_modules/xterm/lib/renderer/Renderer.js.Renderer._renderRows @ Renderer.ts:206
push.../../node_modules/xterm/lib/ui/RenderDebouncer.js.RenderDebouncer._innerRefresh @ RenderDebouncer.ts:56
(anonymous) @ RenderDebouncer.ts:42
requestAnimationFrame (async)
push.../../node_modules/xterm/lib/ui/RenderDebouncer.js.RenderDebouncer.refresh @ RenderDebouncer.ts:42
push.../../node_modules/xterm/lib/renderer/Renderer.js.Renderer.refreshRows @ Renderer.ts:197
push.../../node_modules/xterm/lib/Terminal.js.Terminal.refresh @ Terminal.ts:1204
push.../../node_modules/xterm/lib/Terminal.js.Terminal._onTextAreaBlur @ Terminal.ts:600
(anonymous) @ Terminal.ts:753

```

kittaakos replied:
So, I have checked why it fails on Windows; I do not know why does the `getFileStat` resolve to `null` for an invalid URI on Windows but the URI is incorrect. This is the `toOpen.toString()`.
```
file:///c%3A/Users/kittaakos/Desktop/xxx/c%3A%5CUsers%5Ckittaakos%5CDesktop%5Cxxx
```
My workspace root on Windows is:
```
C:\Users\kittaakos\Desktop\xxx
```

I hope this helps.

AlexTugarev replied:
btw, it's the same on docker/linux.

svenefftinge replied:
Filed https://github.com/theia-ide/theia/issues/5200

svenefftinge commented:
@kittaakos thanks for the review

> The code looks good to me, unfortunately, I could not try it on Windows.

Is that gif from windows, so you tested but it didn't work?

kittaakos commented:
> you tested but it didn't work?

Correct. The exception is also from Windows.

svenefftinge commented on packages/process/src/node/process.ts:
```diff
@@ -172,4 +178,34 @@ export abstract class Process {
+            });
+        } else if (isWindows) {
+            return new Promise<string>(resolve => {
+                resolve(FileUri.create(this.initialCwd).toString());
```
I added the fall back for OSX. injection ApplicationPackage for falling back to `projectPath`  doesn't  seem to be right. As `spawn` without cwd will use the cwd of the parent process if I'm not mistaken. Also it would add another dependency that is never really used as all clients actually pass a cwd. I'd rather make it mandatory.

AlexTugarev commented on packages/terminal/src/browser/terminal-linkmatcher-files.ts:
```diff
@@ -0,0 +1,172 @@
+
+    getValidate(terminalWidget: TerminalWidgetImpl): (link: string) => Promise<boolean> {
+        return async match => {
+            const toOpen = await this.toURI(match, await terminalWidget.cwd);
```
example:
```
<   "//": [
```
`//` is matched as URI by the regex, but cannot be validated as URI internally.

AlexTugarev replied:
matched by this:
```
/(((\.\.?|\~)|([^\0\s!$`&*()\[\]+'":;\\])+)?(\/([^\0\s!$`&*()\[\]+'":;\\])+)+)(((\S*)",[  ]line[  ]((\d+)([  ]column[  ](\d+))?))|((\S*)[  ]on[  ]line[  ]((\d+)(,[  ]column[  ](\d+))?))|((\S*):line[  ]((\d+)(,[  ]column[  ](\d+))?))|(([^\s\(\)]*)(\s?[\(\[](\d+)(,\s?(\d+))?)[\)\]])|(([^:\s\(\)<>'"\[\]]*)(:(\d+))?(:(\d+))?))/g
```

kittaakos replied:
What does this mean, @AlexTugarev?

AlexTugarev replied:
I've seen Theia URI validation errors as result of having the string `//` somewhere in the terminal. Turned out, that it's matched by [`winExcludedPathCharactersClause `](https://github.com/theia-ide/theia/pull/5157/files#diff-28a6b2435812cc9d26163a55fb58d302R153) and passed to `getValidate`. Down the road there is some URI construction and `//` doesn't pass the validation. Sven added a try-catch here.

kittaakos replied:
👍 Thanks! Meanwhile, I have noticed the `Outdated` marker on the comment.

kittaakos commented on packages/terminal/src/browser/terminal-linkmatcher-files.ts:
```diff
@@ -0,0 +1,176 @@
+}
+
+// below reg exps are taken from
+// https://github.com/microsoft/vscode/blob/9edfff8d38d00bd87428af5b9dea7f41a7b5b4d9/src/vs/workbench/contrib/terminal/browser/terminalProcessManager.ts
```
I think the correct link is: https://github.com/microsoft/vscode/blob/fbbc1aa80332189aa0d3006cb2159b79a9eba480/src/vs/workbench/contrib/terminal/browser/terminalLinkHandler.ts

svenefftinge commented:
I have addressed all comments. The windows experience doesn't seem to be good on Vs Code as well as Theia for now. Would be great if someone with a windows machine can have a look. I opened a [separate issue](https://github.com/theia-ide/theia/issues/5200) for that.

kittaakos commented:
> I opened a [separate issue](https://github.com/theia-ide/theia/issues/5200) for that.

👍 

I am trying it once more on macOS.

svenefftinge commented:
> I have noticed one tiny thing, the link is not underlined when it is a diff link, but it works when it is a normal link. See the attached gif. Note; VS Code can underline both.

I think that must have been a rendering glitch. It underlines the diff URLs for me too. Also they are both just matching strategies provided to xterm which takes care of the rendering.
